### PR TITLE
Support for Wireless 3DConnexion SpaceMouse (Pro & Non Pro)

### PR DESCRIPTION
**Pull request by rob-jones**
Added support for 3DConnexion SpaceMouse Pro Wireless. Subclassed vrpn_3DConnexion to provide wireless specific decoding function to cater for differences between 3DConnexion wired and wireless protocols.

### CODE REVIEW

File: vrpn_3DConnexion.C [hunk 1]
```diff
@@ -309,8 +310,115 @@ void vrpn_3DConnexion::decodePacket(size_t bytes, vrpn_uint8 *buffer)
...
+						location = report + 1 + (btn / 8);
+						mask = 1 << (btn % 8);
+						buttons[btn] = ((*location) & mask) != 0;
+					}
+					break;
+
+		case 23:
+			//No need to take action. It just indicates user has stopped using the mouse.
+			break;
+		}
```
Comment (rpavlik) [inline vrpn_3DConnexion.C]: Is this brace in the wrong place? Looks like it opens right after case 3...

File: vrpn_3DConnexion.h [hunk 2]
```diff
@@ -79,6 +79,18 @@ class VRPN_API vrpn_3DConnexion: public vrpn_Button_Filter, public vrpn_Analog {
 };
 #endif  // not VRPN_USE_HID
 
+class VRPN_API vrpn_3DConnexionWireless : public vrpn_3DConnexion {
+public:
+	vrpn_3DConnexionWireless(vrpn_HidAcceptor *filter, unsigned num_buttons,
```
Comment (rpavlik) [inline vrpn_3DConnexion.h]: is `vrpn_HidAcceptor` at least forward-declared even with `VRPN_USE_HID` not declared?

File: vrpn_3DConnexion.C (hunk 1 above)
Comment (rob-jones) [inline vrpn_3DConnexion.C]: Yes Case 23 ended up inside case 3. Correcting now...

File: vrpn_3DConnexion.h (hunk 2 above)
Comment (rob-jones) [inline vrpn_3DConnexion.h]: Prior code to mine reads
Line 27: #if defined(VRPN_USE_HID)
Line 27: ...
Line 30:   vrpn_3DConnexion(vrpn_HidAcceptor *filter, unsigned num_buttons,
Line 52: #else   // not VRPN_USE_HID
line 53: class VRPN_API vrpn_3DConnexion: public vrpn_Button_Filter, public vrpn_Analog {
line 54:   vrpn_3DConnexion(vrpn_HidAcceptor *filter, unsigned num_buttons,

vrpn_HidAcceptor  is used on line 30 when VRPN_USE_HID is declared and line 54 when VRPN_USE_HID is not declared.

I'll admit I've only been able to test this on Windows which has VPRN_USE_HID enabled. I'd rather test it everywhere but not sure I have enough knowledge for that right now.

**Comment by rob-jones** [comment]
Brace placement corrected. Not sure why it didn't show during testing. Still good spot.
Also corrected tabs to spaces to conform with existing style of the files.

**Comment by rpavlik** [comment]
Could just try turning off VRPN use hid in CMake and building again, that would be enough

**Comment by rob-jones** [comment]
I'm using visual studio rather than CMake.
If I comment out the definition of VRPN_USE_HID in vrpn_Configure.h then it still builds but the 3DConnexion mice won't work (which is what I would expect given the code). The non wireless doesn't work or the wireless. My pull request adds support for wireless ontop of where it works for non wireless but I didn't try to add support for using these sorts of mice with other build options. (yes it would be nice to do so but not sure I appreciate all the options for this).

**Comment by rpavlik** [comment]
OK, if it builds with VRPN_USE_HID commented out, then that answers my question.  (You do know you can use CMake to create Visual Studio projects, right? Those are more extensively tested - automatically built with every version of visual studio from 2005 to 2015, and in 64 bit too starting with 2010 I think.)

**Comment by rob-jones** [comment]
>You do know you can use CMake to create Visual Studio projects, right?

I was aware it could, just not familiar enough with the details of how. (I can google it when I have time, no worries).

**Pull request closed without merging**